ledger/complete/wal/checkpoint_v6_leaf_reader.go

```diff
@@ -0,0 +1,106 @@
+package wal
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/common/hash"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/flattener"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/node"
+	"github.com/rs/zerolog"
+)
+
+type LeafNode struct {
+	Hash    hash.Hash
+	Path    ledger.Path
+	Payload *ledger.Payload
+}
+
+type LeafNodeResult struct {
+	LeafNode *LeafNode
+	Err      error
+}
+
+func nodeToLeaf(leaf *node.Node) *LeafNode {
+	return &LeafNode{
+		Hash:    leaf.Hash(),
+		Path:    *leaf.Path(),
+		Payload: leaf.Payload(),
+	}
+}
+
+func OpenAndReadLeafNodesFromCheckpointV6(dir string, fileName string, logger *zerolog.Logger) (
+	allLeafNodesCh <-chan LeafNodeResult, errToReturn error) {
+
+	filepath := filePathCheckpointHeader(dir, fileName)
+
+	f, err := os.Open(filepath)
+	if err != nil {
+		return nil, fmt.Errorf("could not open file %v: %w", filepath, err)
+	}
+	defer func(file *os.File) {
+		errToReturn = closeAndMergeError(file, errToReturn)
+	}(f)
+
+	subtrieChecksums, _, err := readCheckpointHeader(filepath, logger)
+	if err != nil {
+		return nil, fmt.Errorf("could not read header: %w", err)
+	}
+
+	// ensure all checkpoint part file exists, might return os.ErrNotExist error
+	// if a file is missing
+	err = allPartFileExist(dir, fileName, len(subtrieChecksums))
+	if err != nil {
+		return nil, fmt.Errorf("fail to check all checkpoint part file exist: %w", err)
+	}
+
+	bufSize := 1000
+	leafNodesCh := make(chan LeafNodeResult, bufSize)
+	allLeafNodesCh = leafNodesCh
+	defer func() {
+		close(leafNodesCh)
+	}()
+
+	// push leaf nodes to allLeafNodesCh
+	for i, checksum := range subtrieChecksums {
+		readCheckpointSubTrieLeafNodes(leafNodesCh, dir, fileName, i, checksum, logger)
+	}
+
+	return allLeafNodesCh, nil
+}
+
+func readCheckpointSubTrieLeafNodes(leafNodesCh chan<- LeafNodeResult, dir string, fileName string, index int, checksum uint32, logger *zerolog.Logger) {
+	err := processCheckpointSubTrie(dir, fileName, index, checksum, logger,
+		func(reader *Crc32Reader, scratch []byte, nodesCount uint64) error {
```

Comment:
btw, maybe drop the scratch argument from processNode / create the scratch space internally